qualtran/bloqs/chemistry/thc.py

```diff
@@ -0,0 +1,420 @@
+                Register("theta", bitsize=1),
+                Register("succ", bitsize=1),
+                Register("eq_nu_mp1", bitsize=1),
+                Register("plus_a", bitsize=1),
+                Register("plus_b", bitsize=1),
+            ]
+        )
+
+    def __repr__(self) -> str:
+        return (
+            f"PrepareTHC(num_mu={self.num_mu}, num_spin_orb={self.num_spin_orb},"
+            "keep_bitsize={self.keep_bitsize})"
+        )
+
+    @classmethod
+    def build(cls, t_l, zeta, probability_epsilon=1e-8) -> 'PrepareTHC':
+        """Factory method to build PrepareTHC from Hamiltonian coefficients."""
+        assert len(t_l.shape) == 1
+        assert len(zeta.shape) == 2
+        num_mu = zeta.shape[0]
+        num_spat = t_l.shape[0]
+        triu_indices = np.triu_indices(num_mu)
+        num_ut = len(triu_indices[0])
+        flat_data = np.abs(np.concatenate([zeta[triu_indices], t_l]))
+        thetas = [int(t) for t in (1 - np.sign(flat_data)) // 2]
+        alt, keep, mu = preprocess_lcu_coefficients_for_reversible_sampling(
+            flat_data, epsilon=probability_epsilon
+        )
+        num_up_t = len(triu_indices[0])
+        alt_mu = []
+        alt_nu = []
+        alt_theta = []
+        for k in alt:
+            if k < num_up_t:
+                # if k < n * (n + 1) / 2 we are dealing with mu / nu indices
+                alt_mu.append(int(triu_indices[0][k]))
+                alt_nu.append(int(triu_indices[1][k]))
+            else:
+                # else we are dealing with the one-body bit
+                alt_mu.append(int(k - num_ut))
+                alt_nu.append(int(num_mu))
+            alt_theta.append(thetas[k])
+        return PrepareTHC(
+            num_mu,
+            2 * num_spat,
+            alt_mu=tuple(alt_mu),
+            alt_nu=tuple(alt_nu),
+            alt_theta=tuple(alt_theta),
+            theta=tuple(thetas),
+            keep=tuple(keep),
+            keep_bitsize=mu,
+        )
+
+    def build_composite_bloq(self, bb: 'BloqBuilder', **regs: SoquetT) -> Dict[str, 'SoquetT']:
+        data_size = self.num_spin_orb // 2 + self.num_mu * (self.num_mu + 1) // 2
+        log_mu = (self.num_mu).bit_length()
+        log_d = (data_size - 1).bit_length()
+        # Allocate ancillae
+        less_than, alt_theta = bb.split(bb.allocate(2))
+        s = bb.allocate(log_d)
+        sigma = bb.allocate(self.keep_bitsize)
+        keep = bb.allocate(self.keep_bitsize)
+        alt_bitsize = max(max(self.alt_mu).bit_length(), max(self.alt_nu).bit_length())
+        alt_mu = bb.allocate(alt_bitsize)
+        alt_nu = bb.allocate(alt_bitsize)
+        # 1. Prepare THC uniform superposition over mu, nu. succ flags success.
+        mu, nu, succ, eq_nu_mp1 = bb.add(
+            UniformSuperpositionTHC(num_mu=self.num_mu, num_spin_orb=self.num_spin_orb),
+            mu=regs['mu'],
+            nu=regs['nu'],
+            succ=regs['succ'],
+            eq_nu_mp1=regs['eq_nu_mp1'],
```

Comment:
first preference: unpack these in the `build_composite_bloq` method signature

second preference: since you're just passing them through you could probably do `bb.add(UniformSuper, **regs)`

Reply:
Done.